src/requestManager.cpp

```diff
@@ -485,100 +486,62 @@ bool CRequestManager::RequestBlock(CNode *pfrom, CInv obj)
-            // We can only request one graphene block per peer at a time.
-            if (CanGrapheneBlockBeDownloaded(pfrom))
+            // We can only request one thin type block per peer at a time.
+            LOCK(thinrelay.cs_inflight);
+            if (!thinrelay.IsThinTypeBlockInFlight(pfrom, NetMsgType::GRAPHENEBLOCK))
             {
-                // Instead of building a bloom filter here as we would for an xthin, we actually
-                // just need to fill in CMempoolInfo
-                inv2.type = MSG_GRAPHENEBLOCK;
-                CMemPoolInfo receiverMemPoolInfo = GetGrapheneMempoolInfo();
-                ss << inv2;
-                ss << receiverMemPoolInfo;
-                graphenedata.UpdateOutBoundMemPoolInfo(
-                    ::GetSerializeSize(receiverMemPoolInfo, SER_NETWORK, PROTOCOL_VERSION));
-
-                MarkBlockAsInFlight(pfrom->GetId(), obj.hash);
-                AddGrapheneBlockInFlight(pfrom, inv2.hash);
-                pfrom->PushMessage(NetMsgType::GET_GRAPHENE, ss);
-                LOG(GRAPHENE, "Requesting graphene block %s from peer %s\n", inv2.hash.ToString(), pfrom->GetLogName());
-                return true;
-            }
-        }
-        else
-        {
-            // Try to download a graphene block if possible otherwise just download a regular block.
-            // We can only request one graphene block per peer at a time.
-            if (CanGrapheneBlockBeDownloaded(pfrom))
-            {
-                // Instead of building a bloom filter here as we would for an xthin, we actually
-                // just need to fill in CMempoolInfo.
-                inv2.type = MSG_GRAPHENEBLOCK;
-                ss << inv2;
-                ss << GetGrapheneMempoolInfo();
-
-                MarkBlockAsInFlight(pfrom->GetId(), obj.hash);
-                AddGrapheneBlockInFlight(pfrom, inv2.hash);
-                pfrom->PushMessage(NetMsgType::GET_GRAPHENE, ss);
-                LOG(GRAPHENE, "Requesting graphene block %s from peer %s\n", inv2.hash.ToString(), pfrom->GetLogName());
-                return true;
-            }
-            else if (!IsThinBlocksEnabled())
-            {
-                LOG(GRAPHENE, "Requesting regular block %s from peer %s\n", inv2.hash.ToString(), pfrom->GetLogName());
-                std::vector<CInv> vToFetch;
-                inv2.type = MSG_BLOCK;
-                vToFetch.push_back(inv2);
+                // must maintain locking order cs_objDownloder -> cs_inflight, but we also must
+                // ensure that cs_inflight is locked for the duration of both IsThinTypeBlockInFlight()
+                // and AddThinTypeBlockInFlight, or we could end up downloading two of the same block
+                // caused by the multi-threading in message sending and processing.
+                thinrelay.AddThinTypeBlockInFlight(pfrom, inv2.hash, NetMsgType::GRAPHENEBLOCK);
```

Comment:
This requirement for an atomic "if not then add" is very common.  Let's make it one API (how about IfEmptyAddBlockInFlight) that returns true if it was added, false if it already existed.  If you do this, we don't need the LOCK outside the API call, and so we don't need the LEAVE_CRITICAL and ENTER_CRITICAL.  Those 2 APIs should be used as rarely as possible because they are not exception safe and I suspect don't play nicely with the debug mode deadlock detector.

Reply:
yes, thanks, done